extensions/amp-carousel/0.2/amp-carousel.js

```diff
@@ -0,0 +1,160 @@
+import {Carousel} from './carousel.js';
+import {htmlFor} from '../../../src/static-template';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+
+/**
+ * @param {!Element} el The Element to check.
+ * @return {boolean} Whether or not the Element is a sizer Element.
+ */
+function isSizer(el) {
+  return el.tagName == 'I-AMPHTML-SIZER';
+}
+
+class AmpCarousel extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {number} */
+    this.advanceCount_ = 1;
+
+    /** @private {?Carousel} */
+    this.carousel_ = null;
+
+    /** @private {!Array<!Element>} */
+    this.slides_ = [];
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+    const {element} = this;
+    // Grab the slides up front so we can place them later.
+    this.slides_ = Array.from(element.children).filter(c => !isSizer(c));
+    // Create the "Shadow DOM"
+    element.appendChild(this.createDom_());
+
+    this.carousel_ = new Carousel({
+      win,
+      element,
+      scrollContainer: this.element.querySelector('.i-amphtml-carousel-scroll'),
+      runMutate: cb => this.mutateElement(cb),
+    });
+
+    // Handle the initial set of attributes
+    Array.from(this.element.attributes).forEach(attr => {
+      this.attributeChanged_(attr.name, attr.value);
+    });
+
+    this.setupActions_();
+
+    // Do some manual "slot" distribution
+    const scrollContainer = element.querySelector('.i-amphtml-carousel-scroll');
+    this.slides_.forEach(slide => {
+      slide.classList.add('i-amphtml-carousel-slotted');
+      scrollContainer.appendChild(slide);
+    });
+
+    this.carousel_.updateSlides(this.slides_);
+    // Signal for runtime to check children for layout.
+    return this.mutateElement(() => {});
+  }
+
+  /** @override */
+  isRelayoutNeeded() {
+    return true;
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.carousel_.updateUi();
+  }
+
+  /** @override */
+  mutatedAttributesCallback(mutations) {
+    for (const key in mutations) {
+      this.attributeChanged_(key, mutations[key]);
+    }
+  }
+
+  /**
+   * @private
+   */
+  createDom_() {
+    return htmlFor(this.element)`
+      <div class="i-amphtml-carousel-scroll"></div>
+    `;
+  }
+
+  /**
+   * @private
+   */
+  setupActions_() {
+    this.registerAction('prev', () => this.carousel_.prev(), ActionTrust.LOW);
+    this.registerAction('next', () => this.carousel_.next(), ActionTrust.LOW);
+    this.registerAction('goToSlide', ({args}) => {
+      this.carousel_.goToSlide(args['index'] || -1);
+    }, ActionTrust.LOW);
+  }
+
+  /**
+   * @param {string} name The name of the attribute.
+   * @param {string} newValue The new value of the attribute.
+   * @private
+   */
+  attributeChanged_(name, newValue) {
```

Comment:
nit: synonyms are being used: `mutate`, `change` and `update`. I get keeping `update` for the `Carousel` surface, but consider changing `attributeChanged_` to `attributeMutated_`.

Reply:
Done. That seems to fit better with mutatedAttributesCallback. My original thinking was that attributeChanged more closely matched `attributeChangedCallback` from Custom Elements.